File: src/Tests/Behavioral/ArrayPassByValue/ArrayPassByValue.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

internal static void Main() {
    ref var a = ref heap(new array<@string>(2), out var Ꮡa);
    a[0] = "Hello"u8;
    a[1] = "World"u8;
    var p = Ꮡa.at<@string>(0);
    test(a);
    fmt.Println(a[0], a[1]);
    fmt.Println();
    a[0] = "Hello"u8;
    test2(Ꮡa);
    fmt.Println(a[0], a[1]);
    fmt.Println();
    a[0] = "Hello"u8;
    test3(a[..]);
    fmt.Println(a[0], a[1]);
    fmt.Println();
    var primes = new nint[]{2, 3, 5, 7, 11, 13}.array();
    fmt.Println(primes);
    fmt.Println(a[0]);
    stest(p);
    fmt.Println(a[0]);
    assignCopies();
}

internal static array<nint> garr = new nint[]{1, 2, 3}.array();

[GoType] partial struct arrHolder {
    internal array<nint> arr = new(3);
}

internal static void assignCopies() {
    ref var src = ref heap<array<nint>>(out var Ꮡsrc);
    src = new nint[]{1, 2, 3}.array();
    var cpy = src.Clone();
    cpy[0] = 99;
    fmt.Println(src[0], cpy[0]);
    var d = garr.Clone();
    d[0] = 77;
    fmt.Println(garr[0], d[0]);
    array<nint> e = garr.Clone();
    e[1] = 88;
    fmt.Println(garr[1], e[1]);
    e = src.Clone();
    e[2] = 66;
    fmt.Println(src[2], e[2]);
    var h = new arrHolder(arr: new nint[]{4, 5, 6}.array());
    var f = h.arr.Clone();
    f[0] = 55;
    fmt.Println(h.arr[0], f[0]);
    h.arr = src.Clone();
    h.arr[1] = 44;
    fmt.Println(src[1], h.arr[1]);
    var m = new array<nint>[]{new nint[]{7, 8, 9}.array(), new nint[]{10, 11, 12}.array()}.array();
    var row = m[1].Clone();
    row[0] = 33;
    fmt.Println(m[1][0], row[0]);
    var q = Ꮡsrc;
    var g = q.Value.Clone();
    g[0] = 22;
    fmt.Println(src[0], g[0]);
    var (x, y) = (new nint[]{1, 2}.array(), new nint[]{3, 4}.array());
    (x, y) = (y.Clone(), x.Clone());
    x[0] = 11;
    fmt.Println(x[0], x[1], y[0], y[1]);
}

internal static void stest(ж<@string> Ꮡp) {
    ref var p = ref Ꮡp.Value;

    p = "hello"u8;
}

internal static void test(array<@string> a) {
    a = a.Clone();

    fmt.Println(a[0], a[1]);
    a[0] = "Goodbye"u8;
    fmt.Println(a[0], a[1]);
}

internal static void test2(ж<array<@string>> Ꮡa) {
    ref var a = ref Ꮡa.Value;

    fmt.Println(a[0], a[1]);
    a[0] = "Goodbye"u8;
    fmt.Println(a[0], a[1]);
}

internal static void test3(slice<@string> a) {
    fmt.Println(a[0], a[1]);
    a[0] = "Goodbye"u8;
    fmt.Println(a[0], a[1]);
}

} // end main_package

File: src/Tests/Behavioral/ArrayPassByValue/ArrayPassByValue.go
package main

import "fmt"

func main() {
	var a [2]string

	a[0] = "Hello"
	a[1] = "World"

	p := &a[0]

	test(a)
	fmt.Println(a[0], a[1])
	fmt.Println()

	a[0] = "Hello"
	test2(&a)
	fmt.Println(a[0], a[1])
	fmt.Println()

	a[0] = "Hello"
	test3(a[:])
	fmt.Println(a[0], a[1])
	fmt.Println()

	primes := [6]int{2, 3, 5, 7, 11, 13}
	fmt.Println(primes)

	fmt.Println(a[0])
	stest(p)
	fmt.Println(a[0])

	assignCopies()
}

var garr = [3]int{1, 2, 3}

type arrHolder struct {
	arr [3]int
}

// Go array ASSIGNMENT also copies the whole array — a write through the copy
// must never reach the source (locals, globals, fields, elements, derefs).
func assignCopies() {
	// Local := local
	src := [3]int{1, 2, 3}
	cpy := src
	cpy[0] = 99
	fmt.Println(src[0], cpy[0])

	// Local := global
	d := garr
	d[0] = 77
	fmt.Println(garr[0], d[0])

	// var form
	var e = garr
	e[1] = 88
	fmt.Println(garr[1], e[1])

	// Plain reassignment over an existing variable
	e = src
	e[2] = 66
	fmt.Println(src[2], e[2])

	// From a struct field (selector RHS)
	h := arrHolder{arr: [3]int{4, 5, 6}}
	f := h.arr
	f[0] = 55
	fmt.Println(h.arr[0], f[0])

	// Into a struct field (selector LHS)
	h.arr = src
	h.arr[1] = 44
	fmt.Println(src[1], h.arr[1])

	// From an element of an array of arrays (index RHS)
	m := [2][3]int{{7, 8, 9}, {10, 11, 12}}
	row := m[1]
	row[0] = 33
	fmt.Println(m[1][0], row[0])

	// Through a pointer deref (star RHS)
	q := &src
	g := *q
	g[0] = 22
	fmt.Println(src[0], g[0])

	// Swap two arrays by value
	x, y := [2]int{1, 2}, [2]int{3, 4}
	x, y = y, x
	x[0] = 11
	fmt.Println(x[0], x[1], y[0], y[1])
}

func stest(p *string) {
	*p = "hello"
}

// Arrays are passed by value (a full copy)
func test(a [2]string) {
	// Update to array will be local
	fmt.Println(a[0], a[1])
	a[0] = "Goodbye"
	fmt.Println(a[0], a[1])
}

func test2(a *[2]string) {
	fmt.Println(a[0], a[1])
	a[0] = "Goodbye"
	fmt.Println(a[0], a[1])
}

func test3(a []string) {
	fmt.Println(a[0], a[1])
	a[0] = "Goodbye"
	fmt.Println(a[0], a[1])
}
